src/code.c

```diff
@@ -131,16 +133,22 @@ static Expr * ADDR_EXPR(Expr expr)
 */
 static Stat * ADDR_STAT(Stat stat)
 {
+    GAP_ASSERT(STATE(PtrBody) == PTR_BAG(BODY_FUNC(CURR_FUNC())));
+
     return (Stat *)STATE(PtrBody) + stat / sizeof(Stat);
 }
 
 void WRITE_EXPR(Expr expr, UInt idx, UInt val)
 {
+    GAP_ASSERT(expr/sizeof(Stat) + idx < SIZE_BAG(BODY_FUNC(CURR_FUNC())) / sizeof(Expr));
```

Comment:
Why mix Expr and Stat like that? And why not simply this:
```suggestion
    GAP_ASSERT(expr + idx * sizeof(Expr) < SIZE_BAG(BODY_FUNC(CURR_FUNC())));
```

Reply:
Technically that's different if expr doesn't divide sizeof(Stat), and what I wrote lines up more closely with what the code actually does. The Stat/Expr is just a cut+paste problem.

Reply:
Honestly, I don't see how it lines up more closely with how that code works. I'd say to the contrary that what I wrote corresponds more closely to how it works. My version also will catch the case where due to a bug, `expr` is not a multiple of `sizeof(Expr)` (e.g. because somebody passed in an immediate int expression), which one doesn't see if one divides `expr` by `sizeof(Expr)`.

Reply:
The code for ADDR_EXPR/ADDR_STAT does `(Expr *)STATE(PtrBody) + expr / sizeof(Expr);`, so it always divides by sizeof(Expr), and would work fine if it wasn't a multiple.

If we want to require expr is a multiple of sizeof(Expr), we could of course also check that property.